File: Problems/437.PathSumIII/PathSumIII.java
import java.util.ArrayList;
import java.util.List;

public class PathSumIII {
    private int a = 0;
    public int pathSum(TreeNode root, int sum) {
        pathSumHelper(root, sum);
        return a;
    }

    public List<Integer> pathSumHelper(TreeNode root, int sum){
        List<Integer> ans = new ArrayList<>();
        if(root == null)
            return ans;
        if(root.left == null && root.right == null){
            ans.add(root.val);
            if(root.val == sum)
                a++;
            return ans;
        }else{
            List<Integer> l = pathSumHelper(root.left, sum);
            List<Integer> r = pathSumHelper(root.right, sum);
            for(int i : l){
                if(i+root.val == sum)
                    a++;
                ans.add(i+root.val);
            }
            for(int i : r){
                if(i+root.val == sum)
                    a++;
                ans.add(i+root.val);
            }
            if(root.val == sum)
                a++;
            ans.add(root.val);
        }
        return ans;
    }
}

File: path-sum-iii/path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        self.ans =0
        self.root = root
        self.targetSum = targetSum
        self.dfs(self.root, [])
        return self.ans
    def dfs(self, root, curPaths):
        if root == None:
            return 
        curPaths=[x+root.val for x in curPaths]
        count = curPaths.count(self.targetSum)
        self.ans += count
        if root.val == self.targetSum:
            self.ans += 1
        curPaths.append(root.val)   
        self.dfs(root.left, curPaths[:])
        self.dfs(root.right, curPaths[:])
